src/bin/shard/buddy.rs

```diff
@@ -340,3 +342,141 @@ pub fn find_name_accept_buddy(
+        };
+        log_if_failed(
+            clients
+                .get_self()
+                .send_packet(P_FE2CL_REP_PC_BUDDY_WARP_FAIL, &response),
+        );
+        Err(FFError::build(Severity::Info, msg))
+    };
+
+    if player_skyway_state.is_some() {
+        return invalid_warp(
+            3,
+            format!(
+                "Cannot warp to buddy {} - player is currently on a skyway ride",
+                buddy_uid
+            ),
+        );
+    }
+
+    let search = PlayerSearchQuery::ByUID(buddy_uid);
+    let res = search.execute(state);
+    if res.is_none() {
+        if let Some(login_server) = clients.get_login_server() {
+            let cross_shard_pkt = sP_FE2LS_REQ_BUDDY_WARP_VALIDATION {
+                iWarpRequesterPCID: pc_id,
+                iWarpRequesterPCUID: player.get_uid(),
+                iWarpReceiverPCUID: buddy_uid,
+                iWarpRequesterPayzoneFlag: player_payzone_flag as i8,
+                iWarpRequesterShardID: config_get().shard.shard_id.get(),
+            };
+            log_if_failed(
+                login_server.send_packet(P_FE2LS_REQ_BUDDY_WARP_VALIDATION, &cross_shard_pkt),
+            );
+        } else {
+            return Err(FFError::build(
+                Severity::Warning,
+                "No login server connected for cross-shard buddy warp".to_string(),
+            ));
+        }
+        return Ok(());
+    }
+
+    let buddy_id = res.unwrap();
+    let buddy = state.get_player(buddy_id)?;
+    let buddy_payzone_flag = buddy.flags.payzone_flag;
+    let buddy_skyway_state = buddy.skyway_ride.clone();
+    let buddy_pos = buddy.get_position();
+    let buddy_instance = buddy.instance_id;
+
+    if player_payzone_flag != buddy_payzone_flag {
+        return invalid_warp(
+            3,
+            format!(
+                "Cannot warp to buddy {} - payzone flag mismatch (player: {}, buddy: {})",
+                buddy_uid, player_payzone_flag, buddy_payzone_flag
+            ),
+        );
+    }
+
+    if buddy_skyway_state.is_some() {
+        return invalid_warp(
+            3,
+            format!(
+                "Cannot warp to buddy {} - buddy is currently on a skyway ride",
+                buddy_uid
+            ),
+        );
+    }
+
+    if buddy_instance.map_num != ID_OVERWORLD {
+        return invalid_warp(
+            3,
+            format!(
+                "Cannot warp to buddy {} - buddy is in an instance, map_num: {}",
+                buddy_uid, buddy_instance.map_num
+            ),
+        );
+    }
+
+    let player = state.get_player_mut(pc_id)?;
+    player.set_position(buddy_pos);
+    player.instance_id = InstanceID {
+        map_num: ID_OVERWORLD,
+        channel_num: if player.instance_id.channel_num != buddy_instance.channel_num {
+            buddy_instance.channel_num
+        } else {
+            0
+        },
+        instance_num: buddy_instance.instance_num,
+    };
```

Comment:
I don't think we want to set the channel number to 0. Should just copy the instance ID directly from the buddy.
```suggestion
    player.instance_id = buddy_instance;
```